asyncapi/v2/validation.go

```diff
@@ -0,0 +1,50 @@
+package v2
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/asyncapi/event-gateway/asyncapi"
+	"github.com/asyncapi/event-gateway/proxy"
+	"github.com/xeipuuv/gojsonschema"
+)
+
+func FromDocJSONSchemaMessageValidator(doc asyncapi.Document) (proxy.MessageValidator, error) {
+	channels := doc.ApplicationSubscribableChannels()
+	messageSchemas := make(map[string]gojsonschema.JSONLoader)
+	for _, c := range channels {
+		for _, o := range c.Operations() {
+			if !o.IsApplicationSubscribing() {
+				continue
+			}
+
+			// Assuming there is only one message per operation as per Asyncapi 2.x.x.
+			// See https://github.com/asyncapi/event-gateway/issues/10
+			if len(o.Messages()) > 1 {
+				return nil, fmt.Errorf("can not generate message validation for operation %s. Reason: the operation has more than one message and we can't correlate which one is it", o.ID())
+			}
+
+			if len(o.Messages()) == 0 {
+				return nil, fmt.Errorf("can not generate message validation for operation %s. Reason:. Operation has no message. This is totally unexpected", o.ID())
+			}
```

Comment:
Probably I understand the problem described in the mentioned issue - this https://github.com/asyncapi/spec/issues/458 should help for you. However, can we treat the payloads for all defined messages for operation like `oneOf` for the operation? I mean, changing the shape of payload messages to:

```yaml
oneOf:
  - ... shape of payload from message1
  - ... shape of payload from message2
  - ...
```

and then you can validate all messages, but you haven't info from which message you have payload. Is it or not unacceptable in the solution?

Reply:
Good point. it also makes me realize `operation.Messages()` should check for the `oneOf` as well right?

Just to be aligned: you are suggesting that whenever `operation.Messages()` returns more than 1 message, we create a Schema with `oneOf` containing all of those messages and use it for validating. Am I right?

So `operation.Messages()` will return each message individually, but then for making the JSON Schema validation, all of those will be added into just one Schema with a `oneOf`.

Reply:
Nah, nvm. I'm wrong with my assumption. There is only one message but several possible payloads. Is not the same.

Reply:
> Good point. it also makes me realize operation.Messages() should check for the oneOf as well right?

In AsyncAPI we can have construct like:

```yaml
message: 
  payload: ...
  
# or multiple messages

message:
  oneOf:
     - ...
     - ...
```

so `operation.Messages` should read the `oneOf` keyword.

> Nah, nvm. I'm wrong with my assumption. There is only one message but several possible payloads. Is not the same. I think the way it's done right now it's right, isn't?

Yes, for multiple payloads now is ok, but I thought about merging the payloads from messages by `oneOf` keywords, so if you have this:

```yaml
message:
  oneOf:
     - payload: ... # payload1
     - payload: ... # payload2
```

we can merge to:

```yaml
oneOf:
  - ...payload1 
  - ...payload2
```

and then validate all payload from all messages :)

Reply:
But there is only one message per operation. I'm not sure I'm following you. 
If I'm not wrong, `operation.Messages()` should always return one message (technically an array but with only one message) in `v2.x.x`, right? Meaning, the payload of such message will be a Schema that will contain that `oneOf` already and will always validate accordingly.

Reply:
> If I'm not wrong, operation.Messages() should always return one message (technically an array but with only one message) in v2.x.x, right?

Nope :) In `v2.X.X` you can define multiple messages for given operation by construct:

```yaml
message:
  oneOf:
     - payload: ... # payload1
     - payload: ... # payload2
```

So `operation.Messages()` functions should return the array of messages if there is an `oneOf` keyword, from the [spec](https://github.com/asyncapi/spec/blob/master/spec/asyncapi.md#fixed-fields-7):

> A definition of the message that will be published or received on this channel. oneOf is allowed here to specify multiple messages, however, a message MUST be valid only against one of the referenced message objects.

Check `oneOf` mention :)

> Maybe I understand you wrong.

My idea is based on merging only the payloads, when operation has defined multiple messages (and each message has payload field), so validation would be performed on all payloads, but concatenated with `oneOf` keyword. This concatenation should be done in the code, something like:

```go
if len(o.Messages()) > 1 {
  // create new object with concatenating payloads from messages by `oneOf` keyword
}
```

Reply:
However, we cannot infer which message arrives to the gateway (we don't have such a `messageId` field) so your current solution where you are handling only one message is enough.

Reply:
The current solution works. However is not complete since `oneOf` should take into account (even though not extrictly necessary for the event gateway right now).
As the idea is to eventually move all the `asyncapi` package to the `parser-go` repository, I did a small iteration on this code fixing the `oneOf` stuff.

I also fixed the `additionalProperties` and `additionalItems` fields so they are now variadic (either `false` or `*Schema`) so I needed the first for testing.

Reply:
Btw, all fixes on the document interface are being fixed in the `parser-api` repository here: https://github.com/asyncapi/parser-api/pull/10